supervisor/supervisor.go

```diff
@@ -0,0 +1,314 @@
+	}
+
+	_, err := exec.LookPath(cmd)
+
+	if err != nil {
+		msg := fmt.Sprintf("cannot add job, the system cannot find the given executable \"%s\"", cmd)
+		s.logger.Error(msg)
+		return nil, ErrExecNotFound
+	}
+
+	jobId := uuid.New().String()
+	p := NewProcess(cmd, args)
+	sc := p.Start()
+
+	s.wg.Add(1)
+	go func(id string, p *Process, c <-chan *Status) {
+		defer s.wg.Done()
+		s.run(id, p, sc)
+	}(jobId, p, sc)
+
+	s.logger.Info("Added Job: ",
+		zap.String("id", jobId),
+		zap.String("name", cmd),
+		zap.Strings("args", args))
+
+	s.access.Lock()
+	s.jobs.Store(jobId, p)
+	s.outputChannels[jobId] = newLogWatcher()
+	s.ouputBuffers[jobId] = []LogOutput{}
+	s.access.Unlock()
+
+	status := p.GetStatus()
+
+	return &JobStatus{jobId, status}, nil
+}
+
+// run listens for output from the running job through it's StdOut and StdErr channels.
+// Each line of output is first added to the job's line buffer before it's sent to the log watchers.
+func (s *Supervisor) run(id string, proc *Process, sc <-chan *Status) {
+	go func(id string, p *Process) {
+		for {
+			select {
+			case l := <-p.StdOut():
+				line := LogOutput{StdOut, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case l := <-p.StdErr():
+				line := LogOutput{StdErr, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case <-p.Done():
+				return
+			}
+
+		}
+	}(id, proc)
+
+	fs := <-sc
+
+	if fs.ExitCode != 0 || fs.Error != nil {
+		// process didn't exit successfully. Log the error
+		s.logger.Error("Process didn't exit successfully",
+			zap.String("id", id),
+			zap.Int("exit code", fs.ExitCode),
+			zap.Error(fs.Error))
+	}
+}
+
+// Wait returns the done channel from Process. This channel is closed when the process stops running
+func (s *Supervisor) Wait(id string) (chan struct{}, error) {
+	s.access.Lock()
+	defer s.access.Unlock()
+
+	p, ok := s.jobs.Load(id)
+	job := p.(*Process)
+	if !ok {
+		s.logger.Error("Called wait on a job that does not exist", zap.String("id", id))
+		return nil, ErrJobNotFound
+	}
+
+	return job.Done(), nil
+}
+
+// WaitAll waits for all the processes to finish running
+func (s *Supervisor) WaitAll() {
+	s.wg.Wait()
+}
+
+// Stop stops a job by sending it's process group a SIGTERM signal
+func (s *Supervisor) Stop(id string) (*JobStatus, error) {
+	s.access.Lock()
+	defer s.access.Unlock()
+
+	p, ok := s.jobs.Load(id)
+	job := p.(*Process)
+
+	if !ok {
+		s.logger.Error("Cannot find job with the given id", zap.String("id", id))
+		return nil, ErrJobNotFound
+	}
+
+	if err := job.Stop(true); err != nil {
+		s.logger.Error("Cannot stop process", zap.String("id", id))
+		return nil, errors.Wrap(err, "cannot stop process")
+	}
+
+	// Wait for process to exit completely
+	<-job.Done()
```

Comment:
When Stop is called over API, this will hold the client on the wire which is probably not a great UX.

Reply:
Thanks. This has been fixed. The client will be responsible for calling `GetStatus()` to get the final status.